data/binding/binding.go

```diff
@@ -58,31 +57,43 @@ func (l *listener) DataChanged() {
 }
 
 type base struct {
-	listeners async.Map[DataListener, bool]
+	listenerLock sync.Mutex
+	listeners    []DataListener
 
 	lock sync.RWMutex
 }
 
 // AddListener allows a data listener to be informed of changes to this item.
 func (b *base) AddListener(l DataListener) {
-	b.listeners.Store(l, true)
+	b.listenerLock.Lock()
+	b.listeners = append(b.listeners, l)
+	b.listenerLock.Unlock()
 	queueItem(l.DataChanged)
 }
 
 // RemoveListener should be called if the listener is no longer interested in being informed of data change events.
 func (b *base) RemoveListener(l DataListener) {
-	b.listeners.Delete(l)
+	b.listenerLock.Lock()
+	defer b.listenerLock.Unlock()
+
+	for i, listener := range b.listeners {
+		if listener == l {
+			// Delete without preserving order:
+			lastIndex := len(b.listeners) - 1
+			b.listeners[i] = b.listeners[lastIndex]
+			b.listeners[lastIndex] = nil
+			b.listeners = b.listeners[:lastIndex]
+			return
+		}
+	}
+
 }
 
 func (b *base) trigger() {
-	var listeners []DataListener
-	b.listeners.Range(func(listener DataListener, _ bool) bool {
-		listeners = append(listeners, listener)
-		return true
-	})
-
 	queueItem(func() {
-		for _, listen := range listeners {
+		b.listenerLock.Lock()
+		defer b.listenerLock.Unlock()
+		for _, listen := range b.listeners {
```

Comment:
Worth noting that registering a new listener (or unsubscribing) on the same binding in the DataChanged callback will now deadlock. The two slice copy-and-swap-under-lock technique could be used here if we want to work around this limitation.

Reply:
Are you sure it will deadlock? The queued operations run after each other so I can't quite see why it would.

Reply:
I do however have another potential idea for avoiding the lock entirely. How about queuing all operations on the slice onto the main channel? Triggers are the ones that are the most common of course with a fee adds and very few removes.

Reply:
`listenerLock` is held during the invocation of the listeners, and it's held in Add/Remove listener. So it would deadlock if one of the listeners were to try to unregister itself in its callback, for example. I think you're right that queueing all slice operations to main (effectively making them async/promise-like) would avoid locking entirely. I also think duplicating the listeners slice under lock before invoking the callbacks without the lock would have a negligible perf impact, as long as we don't allocate a new slice for it but re-use one.

Reply:
I kind of feel like having two slices wouldn't be worth it for each and every bind able item someone might want to use. That doesn't scale well. Either allocating a slice copy on each trigger (which creates lots of unnecessary garbage if items update often) or queuing to main when also adding or removing listeners are the best options that I find. The latter might have scalability issues but seems better from a performance standpoint given that all triggered listeners already need to be on main (lower frequency of add/remove of listeners than triggers)?

Reply:
I tried the latter. Seems to work fine and be quite a clean implementation